### PR TITLE
AP_NavEKF2: allow for Z compass error changing per flight (4.0 version)

### PR DESCRIPTION
**Pull request by tridge**
this allows for 3x the learned compass offsets in Z axis, and
automatically adjusts the initial learned EKF magZ correction at the
time that final yaw alignment is done on takeoff.

This works around an issue with changes in the Z magnetic field of the
vehicle between flights
This is the 4.0 version of PR #120

### CODE REVIEW

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp [hunk 1]
```diff
@@ -388,6 +389,17 @@ void NavEKF2_core::FuseMagnetometer()
         MagPred[1] = DCM[1][0]*magN + DCM[1][1]*magE  + DCM[1][2]*magD + magYbias;
         MagPred[2] = DCM[2][0]*magN + DCM[2][1]*magE  + DCM[2][2]*magD + magZbias;
 
+        if (pending_z_mag_correction && frontend->_mag_learn_limit > 0) {
+            pending_z_mag_correction = false;
+            float correction = magDataDelayed.mag[2] - MagPred[2];
```
Comment (Aksellence) [inline libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:394]: Same comments as: https://github.com/matternet/ardupilot/pull/120

Lets define '2' with a const or constexpr etc.

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp [hunk 2]
```diff
@@ -388,6 +389,17 @@ void NavEKF2_core::FuseMagnetometer()
...
+            const float correction_limit_ga = frontend->_mag_learn_limit*0.001*3;
```
Comment (Aksellence) [inline libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:395]: 0.001 and 3 are still magic numbers. Lets make them a constant.

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp [hunk 3]
```diff
@@ -388,6 +389,17 @@ void NavEKF2_core::FuseMagnetometer()
...
+            correction = constrain_float(correction,
+                                         -correction_limit_ga, correction_limit_ga);
+            stateStruct.body_magfield[2] += correction;
```
Comment (Aksellence) [inline libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:398]: 2 needs to be a const.

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp [hunk 4]
```diff
@@ -388,6 +389,17 @@ void NavEKF2_core::FuseMagnetometer()
...
+            gcs().send_text(MAV_SEVERITY_INFO, "EKF2 IMU%u magZ correction %.0f",(unsigned)imu_index, correction*1000);
+            MagPred[2] += correction;
```
Comment (Aksellence) [inline libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:400]: same as above

**Review by Aksellence — APPROVED** [review]
Other than the minor changes requested in the comments, the pr looks good.

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:394]: no, the code change follows the coding style for ardupilot. We do not want the matternet EKF2 code to diverge from ardupilot code

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:395]: no, see above

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:400]: no, see above

File: libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp:398]: no, it doesn't. Please don't ask for changes in code style away from how it is done in the upstream ardupilot code. I am working to improve the code style there, but diverging within matternet would be a very bad idea

**Comment by tridge** [comment]
rebased

**Comment by zalice** [comment]
Merged in https://github.com/matternet/ardupilot/pull/159/commits/013491250a6686c16e1c3377f114885eda308972

**Pull request closed without merging**